src/main/java/org/nuxeo/ecm/core/storage/marklogic/CursorService.java

```diff
@@ -0,0 +1,228 @@
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *  
+ * Contributors:
+ *     Kevin Leturc
+ */
+package org.nuxeo.ecm.core.storage.marklogic;
+
+import static org.nuxeo.ecm.core.api.ScrollResultImpl.emptyResult;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.ScrollResult;
+import org.nuxeo.ecm.core.api.ScrollResultImpl;
+
+/**
+ * A cursor service which holds cursors on DB in order to perform scroll operations.
+ *
+ * @param <C> The cursor type.
+ * @param <O> The cursor item type.
+ * @since 9.1
+ */
+public class CursorService<C, O> {
+
+    private static final Log log = LogFactory.getLog(CursorService.class);
+
+    protected Map<String, CursorResult<C, O>> cursorResults = new ConcurrentHashMap<>();
+
+    public void checkForTimedOutScroll() {
+        cursorResults.entrySet().stream().forEach(e -> isScrollTimedOut(e.getKey(), e.getValue()));
+    }
+
+    protected boolean isScrollTimedOut(String scrollId, CursorResult<C, O> cursorResult) {
+        if (cursorResult == null || cursorResult.timedOut()) {
+            if (unregisterCursor(scrollId)) {
+                log.warn("Scroll '" + scrollId + "' timed out");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Registers the input {@link C} and generates a new <code>scrollId</code> to associate with.
+     *
+     * @return the scrollId associated to the cursor.
+     */
+    public String registerCursor(C cursor, int batchSize, int keepAliveSeconds) {
+        return registerCursorResult(new CursorResult<>(cursor, batchSize, keepAliveSeconds));
+    }
+
+    /**
+     * Registers the input {@link C} associated to the input <code>scrollId</code>.
+     *
+     * @return the scrollId associated to the cursor.
+     */
+    public String registerCursor(String scrollId, C cursor, int batchSize, int keepAliveSeconds) {
+        return registerCursorResult(scrollId, new CursorResult<>(cursor, batchSize, keepAliveSeconds));
+    }
+
+    /**
+     * Registers the input {@link CursorResult} and generates a new <code>scrollId</code> to associate with.
+     *
+     * @return the scrollId associated to the cursor result.
+     */
+    public String registerCursorResult(CursorResult<C, O> cursorResult) {
+        String scrollId = UUID.randomUUID().toString();
```

Comment:
If all you want is a random string you can probably use something faster than `UUID`.